include/dice/sparse_map/internal/sparse_hash.hpp

```diff
@@ -0,0 +1,943 @@
+												 typename KeyValueSelect::both_type &>;
+			using pointer = std::conditional_t<IsConst,
+											   typename std::allocator_traits<Allocator>::template rebind_traits<typename KeyValueSelect::both_type>::const_pointer,
+											   typename std::allocator_traits<Allocator>::template rebind_traits<typename KeyValueSelect::both_type>::pointer>;
+
+			// Copy constructor from iterator to const_iterator.
+			sparse_iterator(sparse_iterator<!IsConst> const &other) noexcept requires (IsConst) : cur_bucket_{other.cur_bucket_},
+				  end_bucket_{other.end_bucket_},
+				  bucket_it_{other.bucket_it_} {
+			}
+
+			sparse_iterator(sparse_iterator const &other) noexcept = default;
+			sparse_iterator(sparse_iterator &&other) noexcept = default;
+			sparse_iterator &operator=(sparse_iterator const &other) noexcept = default;
+			sparse_iterator &operator=(sparse_iterator &&other) noexcept = default;
+
+			reference operator*() const noexcept { return KeyValueSelect::both(*bucket_it_); }
+			pointer operator->() const noexcept { return &KeyValueSelect::both(*bucket_it_); }
+
+			sparse_iterator &operator++() noexcept {
+				assert(bucket_it_ != nullptr);
+				++bucket_it_;
+
+				if (bucket_it_ != (*cur_bucket_).end()) {
+					return *this;
+				}
+
+				do {
+					if (++cur_bucket_ == end_bucket_) {
+						bucket_it_ = nullptr;
+						return *this;
+					}
+				} while ((*cur_bucket_).empty());
+
+				bucket_it_ = (*cur_bucket_).begin();
+				return *this;
+			}
+
+			sparse_iterator operator++(int) noexcept {
+				auto tmp = *this;
+				++*this;
+				return tmp;
+			}
+
+			template<bool OIsConst>
+			bool operator==(sparse_iterator<OIsConst> const &other) const noexcept {
+				return cur_bucket_ == other.cur_bucket_ && bucket_it_ == other.bucket_it_;
+			}
+
+			template<bool OIsConst>
+			bool operator!=(sparse_iterator<OIsConst> const &other) const noexcept {
+				return cur_bucket_ != other.cur_bucket_ || bucket_it_ != other.bucket_it_;
+			}
+
+			template<bool OIsConst>
+			bool operator<(sparse_iterator<OIsConst> const &other) const noexcept {
+				return cur_bucket_ < other.cur_bucket_ || (cur_bucket_ == other.cur_bucket_ && bucket_it_ < other.bucket_it_);
+			}
+		};
+
+		iterator mutable_iterator(const_iterator pos) noexcept {
+			// SAFETY: this is non-const therefore the underlying buckets are also non-const
+			// as evidenced by the fact that we can call begin on them
+			auto it_sparse_buckets = buckets_.begin() + std::distance(buckets_.cbegin(), pos.cur_bucket_);
+
+			// SAFETY: this is non-const therefore the underlying sparse array is also non-const
+			auto it_array = sparse_bucket_type::unsafe_mutable_iterator(pos.bucket_it_);
+
+			return iterator{it_sparse_buckets, buckets_.end(), it_array};
+		}
+
+	public:
+		sparse_hash(size_type bucket_count,
+					hasher const &hash,
+					key_equal const &equal,
+					allocator_type const &alloc) : h_{hash},
+												   keq_{equal},
```

Comment:
Does no_unique_address + this result in a no-op?

Reply:
I think this is optimizer dependent. In theory the ctor could still do something, even if it isn't initializing the object.